[inductor] show performance for each autotune config for a kernel

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96249
* __->__ #96248

Be able to benchmark the perf for each config of each kernel.

To use it:
1. run the model with `TORCHINDUCTOR_BENCHMARK_KERNEL` enabled. e.g.:
```
TORCHINDUCTOR_BENCHMARK_KERNEL=1 python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --dashboard --only vgg16 --disable-cudagraphs --training
```
Get the path to the compiled module from log, e.g. 
```
Compiled module path: /tmp/torchinductor_shunting/mj/cmjv5hyt3uq2v7beqkthcl4ul6fh2luwfzmd4tnrquworcmqz4i3.py
```

2. run the compiled module directly with the following options:
- `-k` to benchmark each kernel
- `-c` to benchmark each config for each kernel


Example command:
```
TORCHINDUCTOR_BENCHMARK_KERNEL=1 python /tmp/torchinductor_shunting/mj/cmjv5hyt3uq2v7beqkthcl4ul6fh2luwfzmd4tnrquworcmqz4i3.py -kc
```

Sample result:
<img width="829" alt="Screenshot 2023-03-06 at 6 05 23 PM" src="https://user-images.githubusercontent.com/52589240/223300934-59a4634b-dfd1-46f5-b964-dc0074535236.png">



cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

shunting314 committed:
[inductor] show performance for each autotune config for a kernel

[ghstack-poisoned]

ngimel commented:
Do we need a step to first output compiled module? Can we output results of all autotuning configs as we are autotuning them for the first time, running the model?

shunting314 commented:
> Do we need a step to first output compiled module? Can we output results of all autotuning configs as we are autotuning them for the first time, running the model?

yea, currently we split the following steps
1. generate compiled modules
2. benchmark each kernel in the compiled modules

This way, we can do step 1 once and do step 2 multiple times as we tune hueristics.

If we want, we can also make step 2 being done on the fly while we do step 1 as you mentioned. There is one tricky part here though. We have cache for autotuning result. We would want to ignore the cache if we want to show perf for each config. But  each model is usually being run multiple times in our scripts (for warm up or for more stable perf number), we need
- either take care to only print the autotuning results for the first run. 
- or even better, we only disable autotuning cache for the first run and future run will not even do autotuning because of cache hit

Do we want to go this route?

Chillee commented on torch/_inductor/utils.py:
```diff
@@ -628,18 +628,34 @@ def benchmark_all_kernels(benchmark_name):
-
-        if ms > 0.012 and gb_per_s < 650:
-            print(colorama.Fore.RED + info_str + colorama.Fore.RESET)
+        def get_info_str(ms, prefix=""):
```
btw I'm changing this code a bit to put the kernel name at the end: https://github.com/pytorch/pytorch/pull/96170

Chillee commented on torch/_inductor/config.py:
```diff
@@ -100,6 +100,11 @@ def is_fbcode():
         else os.cpu_count(),
     )
 )
+# be able to override compile_threads. We may want to disable async compiling by
```
Merge this with the above line (perhaps a helper function)? A bit confusing to set this variable twice.

shunting314 committed:
Update on "[inductor] show performance for each autotune config for a kernel"


Be able to benchmark the perf for each config of each kernel.

To use it:
1. run the model with `TORCHINDUCTOR_BENCHMARK_KERNEL` enabled. e.g.:
```
TORCHINDUCTOR_BENCHMARK_KERNEL=1 python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --dashboard --only vgg16 --disable-cudagraphs --training
```
Get the path to the compiled module from log, e.g. 
```
Compiled module path: /tmp/torchinductor_shunting/mj/cmjv5hyt3uq2v7beqkthcl4ul6fh2luwfzmd4tnrquworcmqz4i3.py
```

2. run the compiled module directly with the following options:
- `-k` to benchmark each kernel
- `-c` to benchmark each config for each kernel


Example command:
```
TORCHINDUCTOR_BENCHMARK_KERNEL=1 python /tmp/torchinductor_shunting/mj/cmjv5hyt3uq2v7beqkthcl4ul6fh2luwfzmd4tnrquworcmqz4i3.py -kc
```

Sample result:
<img width="829" alt="Screenshot 2023-03-06 at 6 05 23 PM" src="https://user-images.githubusercontent.com/52589240/223300934-59a4634b-dfd1-46f5-b964-dc0074535236.png">



cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ngimel commented:
Yeah I think two-step process is fine for now, but for the future the disabling cache for the first run that you described would add nice convenience.
The reason I want regular runs to be able to print stats is because sometimes the produced ouput_code is non-runnable or gives skewed performance because of the random inputs, and it's nice to be able to run on "real" inputs.

shunting314 commented:
> Yeah I think two-step process is fine for now, but for the future the disabling cache for the first run that you described would add nice convenience. The reason I want regular runs to be able to print stats is because sometimes the produced ouput_code is non-runnable or gives skewed performance because of the random inputs, and it's nice to be able to run on "real" inputs.

Make sense. Also you pointed out earlier that there may be some alias between inputs that the randomly generated inputs may not be able to capture. We can improve these for sure if they results into problems.

shunting314 commented:
@pytorch merge -f "the test_tensorboard failure is unrelated"

shunting314 commented:
@pytorchbot merge
